cpp/src/lists/sequences.cu

```diff
@@ -0,0 +1,205 @@
+                                        std::optional<column_view> const& steps,
+                                        offset_type const* offsets,
+                                        size_type const* labels,
+                                        rmm::cuda_stream_view stream,
+                                        rmm::mr::device_memory_resource* mr)
+  {
+    auto result =
+      make_fixed_width_column(starts.type(), n_elements, mask_state::UNALLOCATED, stream, mr);
+    if (starts.is_empty()) { return result; }
+
+    auto const result_begin = result->mutable_view().template begin<T>();
+
+    // Use pointers instead of column_device_view to access start and step values should be enough.
+    // This is because we don't need to check for nulls and only support numeric and duration types.
+    auto const starts_begin = starts.template begin<T>();
+    auto const steps_begin  = steps ? steps.value().template begin<T>() : nullptr;
+
+    auto const op = tabulator<T>{starts_begin, steps_begin, offsets, labels};
+    thrust::tabulate(rmm::exec_policy(stream), result_begin, result_begin + n_elements, op);
+
+    return result;
+  }
+};
+
+}  // anonymous namespace
+
+std::unique_ptr<column> sequences(column_view const& starts,
+                                  std::optional<column_view> const& steps,
+                                  column_view const& sizes,
+                                  rmm::cuda_stream_view stream,
+                                  rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(!starts.has_nulls() && !sizes.has_nulls(),
+               "starts and sizes input columns must not have nulls.");
+  CUDF_EXPECTS(starts.size() == sizes.size(),
+               "starts and sizes input columns must have the same number of rows.");
+  CUDF_EXPECTS(cudf::is_index_type(sizes.type()), "Input sizes column must be of integer types.");
+
+  if (steps) {
+    auto const& steps_cv = steps.value();
+    CUDF_EXPECTS(!steps_cv.has_nulls(), "steps input column must not have nulls.");
+    CUDF_EXPECTS(starts.size() == steps_cv.size(),
+                 "starts and steps input columns must have the same number of rows.");
+    CUDF_EXPECTS(starts.type() == steps_cv.type(),
+                 "starts and steps input columns must have the same type.");
+  }
+
+  auto const n_lists = starts.size();
+
+  // Generate list offsets for the output.
+  auto list_offsets = make_numeric_column(
+    data_type(type_to_id<offset_type>()), n_lists + 1, mask_state::UNALLOCATED, stream, mr);
+  auto const offsets_begin  = list_offsets->mutable_view().template begin<offset_type>();
+  auto const sizes_input_it = cudf::detail::indexalator_factory::make_input_iterator(sizes);
+
+  CUDA_TRY(cudaMemsetAsync(offsets_begin, 0, sizeof(offset_type), stream.value()));
+  thrust::inclusive_scan(
+    rmm::exec_policy(stream), sizes_input_it, sizes_input_it + n_lists, offsets_begin + 1);
```

Comment:
Might be a little more clear here to use an exclusive_scan, from 0 to n_lists+1 passing a transform iterator that returns 0 for the "last" element.

Reply:
I just try it. Actually, it is not cleaner because the range of input `sizes` will become `[0, n_lists + 1)` instead of `[0, n_lists)` and thus it will crash if the input is empty (i.e. `n_lists == 0`). In order to handle that crash, we have to add even more code.

Reply:
Are we sure all the code past this point properly handles empty inputs? That's great if so. That said, typically we catch empty inputs early on and special case to do something like `return empty_like`.

Reply:
We return a lists column so no `empty_like` can be used here.

Reply:
Empty input can be handled below that point.
Actually, using `inclusive_scan` is not very bad, just add 2 conditions to check if `n_lists == 0`. I'll push a change for it.